From 5b6b59b2f6249d351a15d5c4c626026c3d174c30 Mon Sep 17 00:00:00 2001
From: Witold Sadowski <wsadowski@marvell.com>
Date: Wed, 17 Mar 2021 07:26:03 -0700
Subject: [PATCH 1334/1921] spi: Fix possible issue in xSPI driver

When xSPI will be configured by probe, and before any operation
on spi memory SMC call will be executed - it will cause an error, as
xSPI driver will not check internal state of IP core before issuing
STIG command.
Fix issue - verify if xSPI is in STIG mode, and run config again if
it is not.

Change-Id: I03a8eab23f0f67b977109453c47f1532c3175321
Signed-off-by: Witold Sadowski <wsadowski@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/47937
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/48123
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/spi/spi-cadence-xspi.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/drivers/spi/spi-cadence-xspi.c b/drivers/spi/spi-cadence-xspi.c
index 4bed78952be7..fda1b3011158 100644
--- a/drivers/spi/spi-cadence-xspi.c
+++ b/drivers/spi/spi-cadence-xspi.c
@@ -303,13 +303,28 @@ bool cdns_xspi_sdma_ready(struct cdns_xspi_dev *cdns_xspi)
 	return true;
 }
 
+static int cdns_verify_stig_mode_config(struct cdns_xspi_dev *cdns_xspi)
+{
+	int cntrl = readl(cdns_xspi->iobase + CDNS_XSPI_CTRL_CONFIG_REG);
+
+	if (FIELD_GET(CDNS_XSPI_CTRL_WORK_MODE, cntrl) != CDNS_XSPI_CTRL_WORK_MODE_STIG)
+		return cdns_xspi_controller_init(cdns_xspi);
+
+	return 0;
+}
+
 static int cdns_xspi_send_stig_command(struct cdns_xspi_dev *cdns_xspi,
 	const struct spi_mem_op *op, bool data_phase)
 {
+	struct device *dev = cdns_xspi->dev;
 	u32 cmd_regs[5] = {0};
 	u32 cmd_status;
 
 	cdns_xspi_wait_for_controller_idle(cdns_xspi);
+	if (cdns_verify_stig_mode_config(cdns_xspi) < 0) {
+		dev_err(dev, "Failed to enter STIG mode\n");
+		return -EPROTO;
+	}
 	cdns_xspi_set_interrupts(cdns_xspi, true);
 	cdns_xspi->sdma_error = false;
 
-- 
2.31.1

